LibTest/js_interop/Float32ListToJSFloat32Array/toJS_t01.dart

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion JSFloat32Array get toJS
+/// Converts this [Float32List] to a [JSFloat32Array] by either casting,
+/// unwrapping, or cloning the [Float32List].
+///
+/// > [!NOTE]
+/// > Depending on whether code is compiled to JavaScript or Wasm, this
+/// > conversion will have different semantics.
+/// > When compiling to JavaScript, all typed lists are the equivalent
+/// > JavaScript typed arrays, and therefore this method simply casts.
+/// > When compiling to Wasm, this [Float32List] may or may not be a wrapper
+/// > depending on if it was converted from JavaScript or instantiated in
+/// > Dart. If it's a wrapper, this method unwraps it. If it's instantiated in
+/// > Dart, this method clones this [Float32List]'s values into a new
+/// > [JSFloat32Array].
+/// > Avoid assuming that modifications to this [Float32List] will affect the
+/// > [JSFloat32Array] and vice versa unless it was instantiated in
+/// > JavaScript.
+///
+/// @description Check that this operation converts this [Float32List] to a
+/// [JSFloat32Array]. Test an array instantiated in Dart.
+/// @author sgrekhov22@gmail.com
+
+import 'dart:js_interop';
+import 'dart:js_interop_unsafe';
+import 'dart:typed_data';
+import '../../../Utils/expect.dart';
+import '../js_utils.dart';
+
+main() {
+  Float32List l1 = Float32List(0);
+  JSFloat32Array a1 = l1.toJS;
+  globalContext["a1"] = a1;
+  eval("globalThis.len1 = globalThis.a1.length;");
+  Expect.equals(0, (globalContext["len1"] as JSNumber).toDartInt);
+  if (isJS) {
+    // Casting case. `toDart` returns the same object.
+    Expect.identical(l1, a1.toDart);
+  }
+  if (isWasm) {
+    // The `Float32List` was instantiated in Dart. This is a cloning case.
+    Expect.notIdentical(l1, a1.toDart);
+  }
+
+  Float32List l2 = Float32List.fromList([1, -3.14, 3333.33]);
+  JSFloat32Array a2 = l2.toJS;
+  Expect.listEquals(l2, a2.toDart.toList());
+  l2[2] = 3.14;
+  if (isJS) {
+    // Casting case. `toDart` returns the same object.
+    Expect.listEquals(l2, a2.toDart.toList());
+  }
+  if (isWasm) {
+    // `a2` was cloned. Changes in `l2` don't affect it.
+    var list = a2.toDart.toList();
+    Expect.equals(1, list[0]);
+    Expect.approxEquals(-3.14, list[1]);
+    Expect.approxEquals(3333.33, list[2]);
+  }
```

Comment:
I'd check a few additional cases here and everywhere else:

- A typed array that's converted to a typed list via `toDart` is wrapped/casted. Modifications to one should affect the other.
- A typed list which was converted from a typed array that's then converted to a typed array should be unwrapped/casted (`toDart.toJS`). The resulting typed array should be the same object as the original.

Reply:
https://github.com/dart-lang/co19/pull/3314

Reply:
> But it'll be the same object on dart2js only, right? On dart2wasm the resulting typed array shouldn't be identical to the original object.

If the typed list (or list) is the result of converting a typed array (or array) to Dart, in dart2wasm, we wrap the original typed array (or array). Then, when we convert it back into JS, if it's already a wrapper, we unwrap it. So the result of `a.toDart.toJS` should be the same as `a`.

I've added better documentation to hopefully make this more clear: https://dart-review.googlesource.com/c/sdk/+/450140